### ISSUE TITLE
Using environment variables in data sources not working anymore

### ISSUE THREAD

**Issue by bthdimension** [grafana-operator/grafana-operator-experimental#34]
I'm currently evaluating the experimental Grafana operator for future use in our organization. I stumbled across the following problem:

This documentation describes the use of environment variables in datasources: [using-environment-variables-in-data-sources](https://github.com/grafana-operator/grafana-operator/blob/master/documentation/env_vars.md#using-environment-variables-in-data-sources )

I tried to set an access token as described in the documentation:
```
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDatasource
...
    secureJsonData:
      httpHeaderValue1: Bearer ${ACCESS_TOKEN}
```

Access did not work. Since the Grafana GUI hides the actual value it is using, I also set the name of a header based on the environment variable:
```
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDatasource
...
    jsonData:
      httpHeaderName1: Authorization
      httpHeaderName2: Bearer ${ACCESS_TOKEN}
```

The Grafana GUI now showed me the literal value that was set instead of replacing with the environment variable. I also tried removing the curly braces, but no luck.

Does the new operator somehow escape the dollar sign?

**Comment by pb82** [comment]
Hey  @bthdimension , this is the development repository for the upcoming version of the Operator. We haven't implemented this feature yet for this version, but it is available for the current v4 release (https://github.com/grafana-operator/grafana-operator)

**Comment by bthdimension** [comment]
Hi @pb82, thanks for your reply. So, the feature will definitely be available again in the upcoming operator but hasn't been implemented yet? 
The current release only allows to deploy Grafana in the same namespace as the operator, so it doesn't work for my use case. I'll be waiting for the upcoming release in that case :-)
Thanks!

**Comment by pb82** [comment]
@bthdimension yes, this will definitely be included in v5. In fact, I think it should already work as the environment variables in the datasource CR are resolved by Grafana. So if you mount a secret/configmap with your variables into the Grafana pod using `spec.deployment.spec` of the Grafana CR, it should already work. I can try this too.

**Comment by bthdimension** [comment]
@pb82 Thanks, would be great if you could try as well. For me, it didn't work. I made sure that I could see the environment variable in the pod's terminal but Grafana didn't replace anything.

**Comment by pb82** [comment]
@bthdimension looking at our example, it might be the way you're referencing the env var: https://github.com/grafana-operator/grafana-operator/blob/master/deploy/examples/datasource-env-vars/Prometheus.yaml#L16

Can you try using `$ACCESS_TOKEN` instead of `${ACCESS_TOKEN}`?

**Comment by NissesSenap** [comment]
@bthdimension have you had time to verify the fix that Peter suggested?

**Comment by bthdimension** [comment]
@pb82 and @NissesSenap Thanks for the hint. $ACCESS_TOKEN was actually the first version I tried and since it didn't work, I tried ${ACCESS_TOKEN} instead. Unfortunately, neither worked.

**Comment by weisdd** [comment]
I suspect that grafana can do env expansion only for data source configs provisioned through a configuration file.
Previously, the operator would pass data source configuration through a file. Nowadays, data sources are configured via API. If I'm right about grafana's capabilities, then we have only two options:
- Do env expansion at the operator's side:
  - Doesn't look like a good scenario as anyone who can create a `GrafanaDatasource` resource will be able to dump the values by directing requests to a specially crafted web server. Secrets management would also be inconvenient due to the need to maintain unique names for each instance;
- Move back to data source provisioning through files:
  - Any changes in data sources would require us to recreate a grafana instance, but I guess it's fine as this part of configuration is normally something stable.

**Comment by weisdd** [comment]
Thinking about that more:
Data source provisioning through files would only work for the grafana instances managed by the operator, which is not an option for "managed grafana" offerings. So, I guess we will need to add env expansion at the operator level anyway. The questions are:
  - Whether env expansion should be something that users can disable if they want;
  - Whether env expansion should only be applicable to external grafana;
  - Whether grafana instances fully managed by the operator should have their data sources provisioned through files.

**Comment by NissesSenap** [comment]
Really good find and explanation @weisdd .

I don't think I have thought enough about this but you write the following and I get a bit confused: 

> * Do env expansion at the operator's side:
>   
>   * Doesn't look like a good scenario as anyone who can create a `GrafanaDatasource` resource will be able to dump the values by directing requests to a specially crafted web server. Secrets management would also be inconvenient due to the need to maintain unique names for each instance;

The operator would only expand environment variables that is configured in the grafana instances?
So if the user don't have access to change the URL of the grafana instance I think we would be "safe" on the other hand it would depend on how much access you have in the kubernetes cluster in general.
You can always write a pod config that matches the grafana service and you would get all the data, you could also try to start a pod on the same node as the grafana instance where you want the password and listen on the http traffic on the node assuming that the cluster don't use a service-mesh or something similar to protect it self.

I don't disagree with you I'm just not sure how much we should protect the grafana-operator users in this case.
At the same time I do think products should protect there users up to a certain point.

I have probably missed something obvious so please poke a whole at my initial thinking.

This is a bit related to https://github.com/grafana-operator/grafana-operator-experimental/issues/44 and how much you can configure around the operator to find data that is outside your namespace.

I know I didn't provide any feedback on your question and I think we will have to bring it up on our next meeting.

**Comment by weisdd** [comment]
@NissesSenap

> The operator would only expand environment variables that is configured in the grafana instances?

The operator would expand variables only in certain resources (e.g. data sources, notifications).

Since it's the operator who does the expansion, the secret values need to be known by manager. Potential options:
- secrets are passed directly to the manager container through environment variables:
  - then `os.ExpandEnv` is enough;
- we need to expand specs of data sources / notifications to add references to secrets:
  - the secrets would have a simple structure like `ENV_NAME: value`;
  - we'll need to implement our own logic for the env expansion. I guess we can simply copy some parts of `os.ExpandEnv`.

For the rest, I might be simply overcomplicating the problem with ephemeral security concerns :)

Anyway, yeah, let's have a discussion on that during the next meeting.

**Comment by pb82** [comment]
Great find @weisdd ! This is a bit of an issue because now the environment variables have to be known to the operator and we're no longer expanding environment variables on the Grafana container. The feature would be clearly different from what Grafana offers (expanding env vars known to the Grafana backend) but I guess we can document it.

This is especially important for external Grafana instances, some providers might rely on data source env vars available only in their environment.

@weisdd @NissesSenap `we need to expand specs of data sources / notifications to add references to secrets` - I was thinking about that too, the datasource spec could look like this:

```yaml
spec:
  instanceSelector:
    matchLabels:
      dashboards: "grafana"
  secrets:
    - datasource-envvars
  datasource:
    name: prom
    type: prometheus
    user: $USER
    password: $PASSWORD
    access: proxy
    url: http://prometheus-service:9090
    isDefault: true
    jsonData:
      'tlsSkipVerify': true
      'timeInterval': "5s"
    editable: true
```

The operator would expect the secret `datasource-envvars` to be in the same namespace as the operator (?).

**Comment by weisdd** [comment]
@pb82

> The operator would expect the secret datasource-envvars to be in the same namespace as the operator (?).

I think secrets should rather be in the same namespace as data sources, because that's the way how secrets are normally referred to in different apps.

**Comment by pb82** [comment]
I'm working on this in https://github.com/grafana-operator/grafana-operator-experimental/pull/61

**Comment by weisdd** [comment]
@bthdimension we merged #61 a few days ago. Could you test the code from the master branch to see if everything works for you as expected?

**Comment by pb82** [comment]
This can be closed, the feature has been implemented.
